sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala

```diff
@@ -162,4 +165,113 @@ class CosmosPatchHelper(diagnosticsConfig: DiagnosticsConfig,
+                          new PatchOperationCore[JsonNode](PatchOperationType.SET, s"/${fieldEntry.getKey}", fieldEntry.getValue)
+                  }
+          }
+      }
+
+      cosmosPatchBulkUpdateOperations.toList
+  }
+
+  def patchBulkUpdateItem(
+                             itemToBeUpdatedOpt: Option[ObjectNode],
+                             patchBulkUpdateOperations: List[PatchOperationCore[JsonNode]]): ObjectNode = {
+
+      // Do not do the modification on original objectNode, as it maybe used by other patchBulkUpdate operations
+      val rootNode = itemToBeUpdatedOpt match {
+          case Some(itemToBeUpdate) => Utils.getSimpleObjectMapper.createObjectNode().setAll(itemToBeUpdate)
+          case _ => Utils.getSimpleObjectMapper.createObjectNode()
+      }
+
+      patchBulkUpdateOperations.foreach(patchOperation => {
+          patchBulkUpdateFields(rootNode, patchOperation.getOperationType, patchOperation.getPath, patchOperation.getResource)
+      })
+
+      rootNode
+  }
+
+  private[this] def patchBulkUpdateFields(
+                                            rootNode: ObjectNode,
+                                            patchOperationType: PatchOperationType,
+                                            path: String,
+                                            pathValue: JsonNode): ObjectNode = {
+      patchOperationType match {
+          case PatchOperationType.SET =>
+              // CosmosDb mapping path: /parent/child1/item
+              // Loop through each path component. If the parent node does not exists, create one {}
+              var parentNode: JsonNode = rootNode
+              val pathArray = path.stripPrefix("/").split("/")
+
+              // get or create parent node
+              for (pathIndex <- 0 until pathArray.size - 1) {
+                  parentNode = getOrCreateNextParentNode(parentNode, s"/${pathArray(pathIndex)}")
+              }
+
+              // after looping through the structure, now set the field value
+              val fieldPath = pathArray(pathArray.size - 1)
+              parentNode match {
+                  case parentIsObjectNode: ObjectNode => parentIsObjectNode.set(fieldPath, pathValue)
+                  case parentIsArrayNode: ArrayNode =>
+                      val arrayIndex = Integer.parseInt(fieldPath)
+                      parentIsArrayNode.set(arrayIndex, pathValue)
+                  case _ => throw new RuntimeException(s"Unsupported parent node type ${parentNode.getClass} in patchBulkUpdateFields") // we should never reach here
+              }
+
+              rootNode
+          case _ => throw new RuntimeException(s"PatchOperationType $patchOperationType is not supported for ItemPatchBulkUpdate") // we should have never reach here
+      }
+  }
+
+  private[this] def getOrCreateNextParentNode(parentNode: JsonNode, childPath: String): JsonNode = {
+      val jsonPath = JsonPointer.compile(childPath)
```

Comment:
QUESTION: compile sounds awfully expensive (worth caching?) - but have never used JsonPointer data type.

Reply:
the compile method seems pretty lightweight, so will keep it for now

Reply:
Please add a comment - saves the next person to investigate again :-)

Reply:
updated